woocommerce/woocommerce
# WC Telemetry: add `first_used` and `installation_date` data to WCTracker

### Submission Review Guidelines:




### Changes proposed in this Pull Request:





Part of https://github.com/woocommerce/woocommerce-ios/issues/10406

After the initial implementation of the telemetry endpoint https://github.com/woocommerce/woocommerce/pull/30415, this PR adds tracking for `first_used` and `installation_date` that are sent to WCTracker for a better understanding of the Woo mobile merchants.

#### `first_used`

When saving the mobile usage data for the first time, `first_used` field is set to be the same as `last_used` (the time of the request). For stores that already have the mobile usage data (`last_used` has been set), they won't have a `first_used` date so that we can track the usage more accurately.

#### `installation_date`

This is an optional field since earlier mobile versions don't track and send the installation date. It's set once when the field hasn't been set before, whether the store has existing mobile usage data or not.

💬  @shiki feel free to suggest a different definition based on the business requirements.



### How to test the changes in this Pull Request:



Prerequisite: the site hasn't had any mobile usage data tracked yet (like a brand new site), and supports https so that it can use the [REST API key & secret](https://woocommerce.com/document/woocommerce-rest-api/) to make API requests. I personally use [Local](https://localwp.com/) for development for the ease of adding SSL to the site.

1. Send a POST request to the `/wp-json/wc-telemetry/tracker` endpoint with body: `{ "platform": "ios", "version": "1.2", "installation_date": "2023-08-08T03:48:50Z" }`. The response should have a `200` status, but no response body.
2. Via WP-CLI or similar, execute `WC_Tracker::get_tracking_data()` --> `wc_mobile_usage` should show an entry for `ios`, and the data underneath should include `version: 1.2`, `installation_date: 2023-08-08T03:48:50+00:00`, and the same `first_used` and `last used` dates in ISO 8601 format
    * Example:  `wp eval "var_dump(WC_Tracker::get_tracking_data()['wc_mobile_usage']);"`
3. Send another POST request to the telemetry endpoint with body: `{ "platform": "ios", "version": "1.2", "installation_date": "2023-08-10T03:48:50Z" }` --> in WCTracker from WP-CLI, `wc_mobile_usage` should show an entry for `ios`, and the data underneath should include `version: 1.2`, `installation_date: 2023-08-08T03:48:50+00:00` (same as the previous value), and the `first_used` date should not be changed from the previous value while `last used` should be updated to the latest request time.
4. You can "reset" the mobile usage data for further testing by deleting the relevant option:
    * `wp option delete woocommerce_mobile_app_usage`

Note: If using `wp shell` to test, you may have to close the shell session and re-open if the option value is cached.
